### PR TITLE
Issue with changing limits of the slider item

### PR DESCRIPTION
**Pull request by Ashwolaa**
After its initialization, it was not possible to change limits of the slideritem .

Fix:
'Bounds' keyword is now accepted in opts to change limits

### CODE REVIEW

File: src/pymodaq_gui/parameter/pymodaq_ptypes/slide.py [hunk 1]
```diff
@@ -125,7 +125,7 @@ def makeWidget(self):
             'value': 0, 'min': None, 'max': None,
              'dec': False,
             'siPrefix': False, 'suffix': '', 'decimals': 12,
-            'int': False
+            'int': False,'bounds':None,
```
Comment (seb5g) [inline src/pymodaq_gui/parameter/pymodaq_ptypes/slide.py]: I have the feeling this bounds is just  on old style equivalent of the limits that have been generalized. I would therefore prefer to change any mention to bounds by limits. And to keep it backcompatible, allows bounds in the limits. The misleading probablt comes from the fact that Parameter use limits but the spinbox/slider user bounds as method.

So I would strongly suggest we keep limits for the behaviour of the Parameter and translate that in the right method for the underlying spinbox and slider

File: src/pymodaq_gui/parameter/pymodaq_ptypes/slide.py [hunk 2]
```diff
@@ -134,14 +134,15 @@ def makeWidget(self):
...
             opts['subtype'] = 'linear'
-        defs['bounds'] = [0., float(self.param.value() or 1)]  # max value set to default value when no max given or 1 if no default
-        if 'limits' not in opts:
-            if 'min' in opts:
-                defs['bounds'][0] = opts['min']
-            if 'max' in opts:
-                defs['bounds'][1] = opts['max']
-        else:
-            defs['bounds'] = opts['limits']
+        if defs['bounds'] is None:
```
Comment (seb5g) [inline src/pymodaq_gui/parameter/pymodaq_ptypes/slide.py:144]: see my comment above about using 'limits' for the keyword in the opts dict

File: src/pymodaq_gui/parameter/pymodaq_ptypes/slide.py [hunk 3]
```diff
@@ -158,6 +159,15 @@ def showEditor(self):
...
         self.widget.spinbox.setFocus()     
 
+    def optsChanged(self, param, opts):
+        # Reimplement optsChanged to correctly update spinbox opts when changed
+        try:
+            super().optsChanged(param, opts)
+        except AttributeError:
+            pass
+        
+        if 'bounds' in opts:
```
Comment (seb5g) [inline src/pymodaq_gui/parameter/pymodaq_ptypes/slide.py:176]: same here

**Review by seb5g — CHANGES_REQUESTED** [review]
I've made some tests and because changing the Parameter limits in a general behaviour triggering a special signal we should stick to it. In conclusion:

* use the keywork limits in Parameter (and init it in defs and opts if not specified by the user or if only min and max are used
* use this limits in opts to trigger a change in the slider and spinbox widgets "bounds" method

thxs

**Comment by seb5g** [comment]
to add some more thought, I saw the ParameterManger is only handling special methods for:

* value_changed
* childadded
* param_deleted

the options change is missing. I'm working on it at the moment and will push a PR in parallel from yours

**Comment by Ashwolaa** [comment]
Ok I will look into it and wait for your patch to see if we can avoid my current hack to detect changes in opts.

I do not really mind about `limits `or `bounds`. However, the reason for this choice comes from the SpinBox widget of pyqtgraph. If you look at the init and setOpts, the keyword that it is currently using is `bounds`.
```
def setOpts(self, **opts):
        """Set options affecting the behavior of the SpinBox.
        
        ============== ========================================================================
        **Arguments:**
        bounds         (min,max) Minimum and maximum values allowed in the SpinBox. 
                       Either may be None to leave the value unbounded. By default, values are
                       unbounded.
        suffix         (str) suffix (units) to display after the numerical value. By default,
                       suffix is an empty str.
        siPrefix       (bool) If True, then an SI prefix is automatically prepended
                       to the units and the value is scaled accordingly. For example,
                       if value=0.003 and suffix='V', then the SpinBox will display
                       "300 mV" (but a call to SpinBox.value will still return 0.003). In case
                       the value represents a dimensionless quantity that might span many
                       orders of magnitude, such as a Reynolds number, an SI
                       prefix is allowed with no suffix. Default is False.
        prefix         (str) String to be prepended to the spin box value. Default is an empty string.
        step           (float) The size of a single step. This is used when clicking the up/
                       down arrows, when rolling the mouse wheel, or when pressing 
                       keyboard arrows while the widget has keyboard focus. Note that
                       the interpretation of this value is different when specifying
                       the 'dec' argument. If 'int' is True, 'step' is rounded to the nearest integer.
                       Default is 0.01 if 'int' is False and 1 otherwise.
        dec            (bool) If True, then the step value will be adjusted to match 
                       the current size of the variable (for example, a value of 15
                       might step in increments of 1 whereas a value of 1500 would
                       step in increments of 100). In this case, the 'step' argument
                       is interpreted *relative* to the current value. The most common
                       'step' values when dec=True are 0.1, 0.2, 0.5, and 1.0. Default is
                       False.
        minStep        (float) When dec=True, this specifies the minimum allowable step size.
        int            (bool) If True, the value is forced to integer type.
                       If True, 'step' is rounded to the nearest integer or defaults to 1.
                       Default is False
        finite         (bool) When False and int=False, infinite values (nan, inf, -inf) are
                       permitted. Default is True.
        wrapping       (bool) If True and both bounds are not None, spin box has circular behavior.
        decimals       (int) Number of decimal values to display. Default is 6. 
        format         (str) Formatting string used to generate the text shown. Formatting is
                       done with ``str.format()`` and makes use of several arguments:
                       
                         * *value* - the unscaled value of the spin box
                         * *prefix* - the prefix string
                         * *prefixGap* - a single space if a prefix is present, or an empty
                           string otherwise
                         * *suffix* - the suffix string
                         * *scaledValue* - the scaled value to use when an SI prefix is present
                         * *siPrefix* - the SI prefix string (if any), or an empty string if
                           this feature has been disabled
                         * *suffixGap* - a single space if a suffix is present, or an empty
                           string otherwise.
        regex          (str or RegexObject) Regular expression used to parse the spinbox text.
                       May contain the following group names:
                       
                       * *number* - matches the numerical portion of the string (mandatory)
                       * *siPrefix* - matches the SI prefix string
                       * *suffix* - matches the suffix string
                       
                       Default is defined in ``pyqtgraph.functions.FLOAT_REGEX``.
        evalFunc       (callable) Fucntion that converts a numerical string to a number,
                       preferrably a Decimal instance. This function handles only the numerical
                       of the text; it does not have access to the suffix or SI prefix.
        compactHeight  (bool) if True, then set the maximum height of the spinbox based on the
                       height of its font. This allows more compact packing on platforms with
                       excessive widget decoration. Default is True.
        ============== ========================================================================
```

**Comment by seb5g** [comment]
yes I do agree but this is related to a widget and not a Parameter for which only limits should be used as with any other Parameter.

**Comment by seb5g** [comment]
> Ok I will look into it and wait for your patch to see if we can avoid my current hack to detect changes in opts.

Yes I especially clarify this in the ParameterManager docstring and in the ParameterEx. pyqtgraph makes a difference between  limits, value, name, default on one hand and any other option on the other hand. You'll see in my PR that I therefore also treat it differently. Note: the name change should never be used dynamically (at least in plugins) because it would break the relation between UI and plugin. That's why I don't have a dedicated method in Parameter Manager

**Comment by seb5g** [comment]
My PR #5  has been added to the 5.0.x_dev branch. You can resync and use that for better handling and testing of the slider feature

**Comment by Ashwolaa** [comment]
Reimplementing the limitschanged function gives excellent results. This effectively makes it direct either with the keyword limits.

Status of PR is that both limits and bounds are accepted as keywords. However, consistency is still required. So if you initialize it as limits, you should keep using limits

**Comment by seb5g** [comment]
what if someone uses min and max keyword? or only min? or only max? did you take this into account? Could you build some tests on that to make sure the behaviour is fixed?

**Comment by Ashwolaa** [comment]
> what if someone uses min and max keyword? or only min? or only max? did you take this into account? Could you build some tests on that to make sure the behaviour is fixed?

I looked a bit into it and it gets quite cumbersome in the implementation.
The reason for it is that min/max is not managed at all in the spinbox widget initially. In fact, the current spinbox parameter does not manage min and max currently. What I mean is :
` settings.setOpts(min=10,max=20) ` would not modify the limits of the widget.
while:
` settings.setOpts(bounds=[0,100]) ` or ` settings.setLimits([0,100]) ` would modify it.

So to sum up, I think we should stick to these keywords as you initially asked as a start up.

**Comment by seb5g** [comment]
Be aware that they are two places where options are considered:
* in the parameter
* in the widget

The fact that spinbox is not handling it is why we should handle it with the Parameter object. Whenever the options min , max, limits or bounds are changed, we should modify one or the other accordingly using self.opts[] (self is the parameter) and fire self.LimitsChanged(). Like this it would be ok!

I insist because min/max are also used with the 'float' and 'int' parameter

**Comment by seb5g** [comment]
see the numeric module in pyqtgraph: 
image

**Comment by seb5g** [comment]
be also aware that pyqtgraph usea particular implementation of SpinBox with more options than the standard QSpinBox

**Comment by Ashwolaa** [comment]
> Be aware that they are two places where options are considered:
> 
> * in the parameter
> * in the widget
> 
> The fact that spinbox is not handling it is why we should handle it with the Parameter object. Whenever the options min , max, limits or bounds are changed, we should modify one or the other accordingly using self.opts[] (self is the parameter) and fire self.LimitsChanged(). Like this it would be ok!
> 
> I insist because min/max are also used with the 'float' and 'int' parameter

I agree but this behavior is not being handled with the numeric parameter item as shown with :
`settings.setOpts(min=10,max=20)`

So my point is that this change should be also be adressed for the numeric type so maybe a subclassing of the numeric parameter item is a better approach. We could then use it as a base model for the sliderparameter item.

Are you ok with that?

**Comment by seb5g** [comment]
> I agree but this behavior is not being handled with the numeric parameter item as shown with :
> `settings.setOpts(min=10,max=20)`

I don't understand this point?!

But you are right that the best approach would probably inherit from the Numeric Parameter, and use our slider +spinbox widget as a replacement of the simple spinbox. Be aware that I myself (in pymodaq_gui) subclassed the numeric parameter to change slightly the behavior when setting new limits. So make sure to inherit from my numeric class. cheers

**Comment by seb5g** [comment]
ok I think I got it. Indeed it is actually not implemented in OptsChanged of the Numeric Parameter:
```
    def optsChanged(self, param, opts):
        super().optsChanged(param, opts)
        sbOpts = {}
        if 'units' in opts and 'suffix' not in opts:
            sbOpts['suffix'] = opts['units']
        for k, v in opts.items():
            if k in self.widget.opts:
                sbOpts[k] = v
        self.widget.setOpts(**sbOpts)
        self.updateDisplayLabel()
```

then you could reimplement optsChanged within my subclass of numeric Parameter and then it would be inherited by the slider as well!

**Review by seb5g — APPROVED** [review]
